crates/shim-sgx/src/heap.rs

```diff
@@ -48,38 +54,54 @@ impl Heap {
     }
 
     /// Increase or decrease `brk` address.
-    /// TODO: unmap memory, once support for EMODT is added.
     pub fn brk(&mut self, brk: Address<usize, Page>) -> Address<usize, Page> {
         if brk < self.start || brk >= self.end {
             return self.brk;
         }
-        if brk > self.brk_max {
-            let length = Offset::from_items((brk.raw() - self.brk_max.raw()) / Page::SIZE);
-            let region = Region::new(self.brk_max, self.brk_max + length);
-            if self
-                .ledger
-                .map(region, Access::READ | Access::WRITE)
-                .is_err()
-            {
-                return self.brk;
-            }
-            self.brk_max = brk;
+
+        if brk <= self.brk_max {
+            self.brk = brk;
+            return brk;
         }
-        self.brk = brk;
-        self.brk
+
+        let region = Region::new(self.brk_max, brk);
+
+        if self.ledger.overlaps(region) {
+            return self.brk;
+        }
+
+        if self
+            .ledger
+            .map(region, Access::READ | Access::WRITE)
+            .is_err()
+        {
+            return self.brk;
+        }
+
+        self.brk_max = brk;
+        brk
     }
 
     /// Find and reserve an address range.
     pub fn mmap(
         &mut self,
+        addr: Option<Address<usize, Page>>,
         length: Offset<usize, Page>,
         access: Access,
     ) -> Option<Address<usize, Page>> {
-        let addr = self.ledger.find_free_back(length)?;
+        let addr = match addr {
+            None => self.ledger.find_free_back(length)?,
+            Some(addr) => addr,
+        };
         let region = Region::new(addr, addr + length);
         self.ledger.map(region, access).ok()?;
         Some(addr)
     }
+
+    /// Release a region.
+    pub fn munmap(&mut self, region: Region) {
+        self.ledger.unmap(region).unwrap();
+    }
```

Comment:
Quoting from the `munmap(2)` man page:
>  It is not an error if the indicated range does not contain any mapped pages.

Do we really want to panic here?

Reply:
I rethought the implementation a bit, thanks for the remark.